[KOGITO-7764] Timeout support (#198)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc

= Timeouts on events for {context}
:compat-mode!:
// Metadata:
:description: Using timeouts in {context}
:keywords: kogito, workflow, serverless, timeout, timer, expiration

> When defining a state in a workflow you can set a timeout which is a property used to control the maximum time this state should take to be complete. In case of the time is elapsed and overdue, this state is considered timed-out, and the workflow moves to the next stated defined in the transition. The properties that you can use when defining a state timeout are described in the link:{spec_doc_url}#event-timeout-definition[Serverless Workflow specification].

Reviewer: In any case we must use "timeouts" (with s) instead of timeout

> When defining a state in a workflow you can set a timeout which is a property used to control the maximum time this state should take to be complete. In case of the time is elapsed and overdue, this state is considered timed-out, and the workflow moves to the next stated defined in the transition. The properties that you can use when defining a state timeout are described in the link:{spec_doc_url}#event-timeout-definition[Serverless Workflow specification].

Author: +1 incorporating you suggestion.

> When defining a state in a workflow you can set a timeout which is a property used to control the maximum time this state should take to be complete. In case of the time is elapsed and overdue, this state is considered timed-out, and the workflow moves to the next stated defined in the transition. The properties that you can use when defining a state timeout are described in the link:{spec_doc_url}#event-timeout-definition[Serverless Workflow specification].

Reviewer: thanks

> The event-based states can define a property `eventTimeout` which uses ISO 8601 data and time format to specify a duration of time, for instance, you can if you can represent 15 minutes like `PT15M`,  to represent 15 minutes or 2 days, 3 hours and 4 minutes `P2DT3H4M`.

Reviewer: I believe that here we can maybe include something like this:

"Event-based states can use the sub-property eventTimeout to configure the maximum time to wait for an event to arrive."

Since "timeouts" is a composed property. Another alternative could be to start saying in the initial paragrap that timeouts is a composed thing, but I think it's ok as it is, since we introduce the concept.

Reviewer: thanks

Reviewer: wait!

I believe this piece needs some more love:

"for instance, you can if you can represent 15 minutes like PT15M, to represent 15 minutes or 2 days, 3 hours and 4 minutes P2DT3H4M."

Author: +1, changing the text

[NOTE]
====

> Event timeouts can not be defined as a specific point in time, but it should be an amount of time, a duration, which is considered to start when the referred state becomes active in the workflow.

Reviewer: I believe that if in the previous paragrpah we say that the event timeout must be a duration, we don't need this note.

====

[IMPORTANT]
====

> {product_name} currently, supports timeout only for *Callback* and *Switch* states with events, other states are going to support timeouts in the future releases.

Reviewer: Maybe...

"Current kogito version, has timeouts support only for the Callbakc and Swith states with events. Other states will be included in future releases."

Reviewer: this change was not applied, not mandatory, but has the +1 from your side. Maybe you forgot.

Reviewer: Ah,  see, was applied but you kept "Kogito currently, ....", ok, let's  @hmanwani-rh take a look. for me "Kogito currently, bla,bla " sounds strante.

====

=== Callback state timeout

> Callback state can be used when you need to publish an event to trigger an external service and waits an asynchronous response as an event, the callback. Once this event is consumed, the workflow continues the execution to the next state defined in the transition property. See more on xref:eventing/working-with-callbacks.adoc[Callback state in {context}].

Reviewer: Maybe

"Callback state can be used when you need to execute an action, for example invoke an external service, and wait for an asynchronous response in the form of an event, the callback"

Reviewer: Maybe
"When the event arrives, it is consumed, and the workflow continues...."

Reviewer: thanks

> Since the callback event halts the execution util the event is consumed, you can define an `eventTimeout` for it, and in case the event does not arrive in the defined duration time, the workflow continues the execution moving to the next state defined in the transition, see the <<callback-state, example>>.

Reviewer: thanks

[#callback-state]
.Example of callback state with timeout
[source,json]
----
{
 "name": "CallbackState",
 "type": "callback",
 "action": {
   "name": "callbackAction",
   "functionRef": {
     "refName": "callbackFunction",
     "arguments": {

>        "input": "${\"callback-state-timeouts: \" + $WORKFLOW.instanceId + \" has executed the callbackFunction.\"}"

Reviewer 2: I think this argument can be an object instead. This way, you avoid adding these scape chars and educate the users.

Author: yeah, @wmedvede can you update the example in this way we align with the guide

Reviewer: Hi @ricardozanini, I missed this comment, but what do you mean with an object?

Let me explain:
The idea here is that the invoked function is sysout, why? to make things simple, etc., but also during the review process it was also suggested to remove a kind of "more real service" invocation I had, that in the end did the system out.
So, the idea is, that this two SWs are intended to show specifically how to use the timeouts and to provide a quickly way to see understand the code, execute, and see the results in the system output. Let's say a one-to-one matching.

Not sure if I've got your point, but to have an object as argument, I need to create it first, so, or we pass it to the SW as parameter, or we use an inject state, or we create it somehow in another point of the SW, etc.
But with this options, I loose the end-to-end mapping from the SW definition to the system output, and the simplicity of reading the flow.

What do you suggest?

Reviewer 2: Oh ok, this is used in sysout, I missed that. This is fine.

     }
   }
 },
 "eventRef": "callbackEvent",
 "transition": "CheckEventArrival",
 "onErrors": [
   {
     "errorRef": "callbackError",
     "transition": "FinalizeWithError"
   }
 ],
 "timeouts": {
   "eventTimeout": "PT30S"
 }
}
----

=== Switch state timeout

> Switch states can be defined with link:{spec_doc_url}#switch-state-event-conditions[eventConditions] property, where the workflow execution waits to make a decision depending on the events to be consumed and matched, defined through link:{spec_doc_url}#event-definition[event definition].

Reviewer: Maybe here we can say something like 
"The switch state can be used when you need too..." 
to get it more similar to the previous paragraph, and then the rest

> In this situation, you can define an event timeout, that controls the maximum time to wait for an event to match the conditions, if this time is expired, the workflow moves to the state defined in the `defaultCondition` property in the switch state, as showed in the <<switch-state, example>>.

Reviewer: in the `defaultCondition` of the switch state, as you can see in the following example.

Maybe the link "<<switch-state, example>>." is in needed, since it's in the line below

Reviewer: not sure if this was applied, I tried to say that the link is not needed since the example comes in the next lines. But np.

Author: +1, anyway I think it is great to have the link the text on the link remains the same.

See more details about this state on the link:{spec_doc_url}#switch-date[Serverless Workflow specification].

[#switch-state]
.Example of switch state with timeout
[source,json]
----
{
    "name": "ChooseOnEvent",
    "type": "switch",
    "eventConditions": [
    {
        "eventRef": "visaApprovedEvent",
        "transition": "ApprovedVisa"
    },
    {
        "eventRef": "visaDeniedEvent",
        "transition": "DeniedVisa"
    }
    ],
        "defaultCondition": {
        "transition": "HandleNoVisaDecision"
    },
        "timeouts": {
        "eventTimeout": "PT5S"
    }
}
----

=== Deploying a timed-based workflow

> n order to deploy a work that contains timeouts or any other timer-based action, it is necessary to have Job Service running in your environment, which is an external service responsible for workflows timers, see the <<job-service, section>> for more information.

Reviewer: The paragraph should starts with "i" in lowercase.

I think we should say "you can see the details" insead of "in deatils", maybe @hmanwani-rh can help with the most correct form.

> n order to deploy a work that contains timeouts or any other timer-based action, it is necessary to have Job Service running in your environment, which is an external service responsible for workflows timers, see the <<job-service, section>> for more information.

Author: +1

In the <<timeout-example, timeout example>> you can see in details how to set up a knative infrastructure with the workflow and job service running.

> [#job-service]

Reviewer: Jobs Service or Job Service?, both are good, but I think we don't have a fixed name yet. Sometimes we say job, other times we say jobs.
In the kogito business automation documentation we say "Jobs Service"

Reviewer 2: cc @hmanwani-rh

Reviewer 2: cc @kaldesai now :P

=== Job Service configuration

All timer-related actions that might be declared in a workflow, are handled by a supporting service, called Job Service, which is responsible for managing, scheduling, and firing all actions (jobs) to be executed in the workflows.

Suppose the workflow service is not configured to use job service or there is no such service running. In that case, all timer-related actions use an embedded in-memory implementation of job service, which should not be used in production, since when the application shutdown, all timers are lost, which in a serverless architecture is a very common behavior with the scale to zero approach. That said, the no job service configuration can only be used for testing or development, but not for production.

> The main goal of the Job Service is to work with only active jobs. The Job Service tracks only the jobs that are scheduled and that need to be executed. When a job reaches a final state, the job is removed from the Job Service. All job information and transition states are sent to the Kogito Data Index Service where they can be indexed and made available for GraphQL queries.

Reviewer: When configured, all the jobs information and status changes are sent to the Kogito Data Index Service, where they can be indexed and made available for GraphQL queries.

Reviewer: thanks

==== Job Service persistence

An important configuration aspect of job service is the persistence mechanism, where all job information is stored in a database that makes this information durable upon service restarts and guarantees no information is lost.

==== PostgreSQL

> Using PostgreSQL is the recommended database to be used with job service where it is already configured with Flyway for the schema creation, in this way all tables are created by the service, in case you need to externally control the database schema, you can check the SQL scripts in (link) and apply them.

Reviewer: May be we can rephrase the start of this paragraph to something like this.

PostgreSQL is the recommended database to use with job service.
Additionally, the jobs service provides an initialization procedure that integrates Flyway for the database initialization, in this way, all the tables, etc., are automatically created by the service.

Reviewer: thanks

> You need to set the proper configuration parameters when starting job service. The example covers running PostgreSQL as a Kubernetes deployment but you can run it the way it fits in your environment, the important part is to set all the configuration parameters points to your running instance of PostgreSQL.

Reviewer: Maybe

"The example shows how to run PostgreSQL as a Kubernetes deployment but you can run it the way it fits in your ....."

Reviewer: thanks

Job Service configuration

==== Ephemeral
Alternatively, there is an in-memory database implementation that does not require any external database configuration, similar to the workflow runtime embedded in-memory job service (link), it can be used for testing and development purposes, but it is not recommended for production, since all jobs are lost in case of a service restart or failure.

> === Job Service communication

Reviewer: Don't understood too much this paragraph, but yeah, maybe it's not ready.

Reviewer: thanks

[NOTE]
====
The Job Service does not execute a job but triggers a callback that might be an HTTP request or a Cloud Event that is managed by the configured jobs addon see(link addons config) in the workflow application.
====

==== Knative Eventing

> You can configure job service to consume and publish events through knative setting some configuration parameters in your environment.

Reviewer: Maybe we can start with something like this:

"To rely the communication between the job service and the workflows runtime to the knative eventing system,  you must provide a set of configurations."

"The Job Service configuration is done through the deployment descriptor shown in the example."

> You can configure job service to consume and publish events through knative setting some configuration parameters in your environment.

Reviewer: "is done through an addon, (<-comma) which is responsible for publishing and consuming events related to timers.

> You can configure job service to consume and publish events through knative setting some configuration parameters in your environment.

Reviewer: When you run the workflow as a knative service, you must add "bla bla bla" to your project and provide the proper configuration.

> You can configure job service to consume and publish events through knative setting some configuration parameters in your environment.

Author: +1

Addon configuration
The communication from the workflow application with Job Service is done through an addon which is responsible for publishing and consuming events related to timers. When running the workflow as a knative service, the kogito-addons-quarkus-jobs-knative-eventing should be added to the project as a dependency alongside the proper configuration.

Dependency:
.Callback state example with timeout
[source, xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-quarkus-jobs-knative-eventing</artifactId>
</dependency>
----

Configuration parameters:

.Callback state example with timeout
[source, properties]
----
# Events produced by kogito-addons-quarkus-jobs-knative-eventing to program the timers on the job service.
mp.messaging.outgoing.kogito-job-service-job-request-events.connector=quarkus-http
mp.messaging.outgoing.kogito-job-service-job-request-events.url=${K_SINK:http://localhost:8280/jobs/events}
mp.messaging.outgoing.kogito-job-service-job-request-events.method=POST
----

[#timeout-example]
== Timeout showcase example

> In the link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[serverless-workflow-timeouts-showcase] you can see a complete architecture, containing a serverless workflow application with timeouts configured alongside Job service running on Knative.

Reviewer: Maybe start with something like this:

In the serverless-workflow-timeouts-showcase you can see an end-to-end example that contains....... 
integrates.....

etc

Reviewer: sometimes we write job service, Job Service, Job service. Wondering if we should try to standarize always uppercased "Job Serivce" maybe?

Author: I'm fine, let's keep Job Service and let @hmanwani-rh give her suggestion later on the review.

> There are two workflows to showcase timeouts usage: `Callback` and `Switch` states.

Reviewer: There are two workflows that showcase the timeouts usage in the: Callback and Switch states.

Reviewer: thanks

=== Callback workflow

> It is a simple workflow where once the execution reaches the callback state it waits for the event `callbackEvent`to be received and continue the execution.

Reviewer: It is a simple workflow where, (a comma here I believe)

Reviewer: "event `callbackEvent`to arrive" instead of "be received"

Reviewer: thanks

.Callback timeout workflow
image::core/callback-state-timeouts.svg[]

.Callback event
[source, json]
----
{
"name": "callbackEvent",
"source": "",
"type": "callback_event_type"
}
----

> A timeout is configured with a maximum time 30 seconds to be waited by the workflow to receive the callbackEvent, in case it does not arrive in time, the execution moves, and the eventData variable remains null.

Reviewer: The timeout is configured... 

Maybe to show the callback sate json element instead of the event can help.?

See the <<callback-state, callback state definition>>.

=== Switch workflow

> The switch example is similar to the callback but once the execution reaches the state, it waits for once of the two configured events, `visaDeniedEvent` or `visaApprovedEvent`, to be consumed, see the <<switch-state, switch state definition>>.

Reviewer: it waits for ONE of the two configured events, visaDeniedEvent or visaApprovedEvent, to ARRIVE, see the [switch state definition](https://deploy-preview-198--tangerine-crostata-04a6bf.netlify.app/serverlessworkflow/latest/core/timeouts-support.html#switch-state).

Whenever one of the events is consumed, before the configured timeout, the workflow execution moves to the next state defined in the `transition`.

> If none of the events arrive before 30 seconds of timeout, the workflow then moves to the state defined in `defaultCondition' transition.

Reviewer: Maybe:
If none of the events arrive before the timeout is overdue, the workflow then moves to the state defined in `defaultCondition' transition.

.Switch timeout workflow
image::core/switch-state-timeouts.svg[]

=== Running the example

> In order to run the example you need to have a kubernetes or Openshift cluster running with Knative configured. In the example minikube is being used you can follow the steps described in the example's link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[readme].

Reviewer: To run the example you must have access too a kubernetes....

Reviewer: For simplicity, the example use minikube... bla, bla

Author: +1

> All the descriptor files used to deploy the example infrastructure are present in the kubernetes folder, and for the workflow application, the descriptors are placed on the target/kubernetes, generated after the build.

Reviewer: Maybe here we can add a [NOTE] ?
And spit a bit phrase, like "The descriptor related to the showcase-databse and jobs service deployment are located at....
The descriptors related to the application....

> The diagram below describes how is the architecture of the example to be deployed on kubernetes and Knative.

Reviewer: The following diagram shows the example's architecture when it is deployed in the Kubernetes + Knative infrastructure.

Author: +1

Author: +1

[source, shell]
----
service.serving.knative.dev/timeouts-showcase created

trigger.eventing.knative.dev/visa-denied-event-type-trigger-timeouts-showcase created
trigger.eventing.knative.dev/visa-approved-event-type-trigger-timeouts-showcase created
trigger.eventing.knative.dev/callback-event-type-trigger-timeouts-showcase created
sinkbinding.sources.knative.dev/sb-timeouts-showcase created
----

==== Creating a workflow instance

To create a workflow you can interact with the workflow using the provided REST APIs, in the example provide a test Web UI to make it easy to test.

First, you need to get the service URL on the cluster.

.Getting the workflow service URL on the cluster
[source, shell]
----
kn service list | grep timeouts-showcase
----

.Service URL in the response, similar to this.
[source, shell]
----
NAME                      URL                                                             LATEST                          AGE     CONDITIONS   READY   REASON
timeouts-showcase         http://timeouts-showcase.default.10.105.86.217.sslip.io         timeouts-showcase-00001         3m50s   3 OK / 3     True
----

> === Using the show UI

Reviewer: "Using the showcase" instead of "show"

Author: +1

= Timeouts on events for {context}
:compat-mode!:
// Metadata:
:description: Using timeouts in {context}
:keywords: kogito, workflow, serverless, timeout, timer, expiration

When you define a state in a serverless workflow, you can use the `timeouts` property to configure the maximum time to complete this state.
When that time is overdue, the state is considered timed-out, and the engine continues the execution from this state. The execution flow depends on the state type, for instance,
a transition to a next state.
All the properties you can use to configure state timeouts are described in the link:{spec_doc_url}#event-timeout-definition[Serverless Workflow specification].

Event-based states can use the sub-property `eventTimeout` to configure the maximum time to wait for an event to arrive.

> This property uses `ISO 8601` data and time standard to specify a duration of time. It follows the format `PnDTnHnMn.nS` with days considered to be exactly 24 hours. For instance,

Reviewer 2: ```suggestion
This property uses link:https://en.wikipedia.org/wiki/ISO_8601[`ISO 8601` data and time standard] to specify a duration of time. It follows the format `PnDTnHnMn.nS` with days considered to be exactly 24 hours. For instance,
```

You may consider adding this link as a page attribute.

`PT15M` configures 15 minutes, and `P2DT3H4M` defines 2 days, 3 hours and 4 minutes.

[NOTE]
====
Event timeouts can not be defined as a specific point in time, but it should be an amount of time, a duration, which is considered to start when the referred state becomes active in the workflow.
====

[IMPORTANT]
====

> {product_name} currently, has timeouts support only for *Callback* and *Switch* states with events. Other states will be included in the future releases.

Reviewer 2: Please add this information to the CNCF guide.

Author: +1

In the <<timeout-example, timeout example>> you can see the details of how set up a knative infrastructure with the workflow and job service running.

[#job-service]
=== Job Service configuration

All timer-related actions that might be declared in a workflow, are handled by a supporting service, called Job Service, which is responsible for managing, scheduling, and firing all actions (jobs) to be executed in the workflows.

Suppose the workflow service is not configured to use job service or there is no such service running. In that case, all timer-related actions use an embedded in-memory implementation of job service, which should not be used in production, since when the application shutdown, all timers are lost, which in a serverless architecture is a very common behavior with the scale to zero approach. That said, the no job service configuration can only be used for testing or development, but not for production.

The main goal of the Job Service is to work with only active jobs. The Job Service tracks only the jobs that are scheduled and that need to be executed. When a job reaches a final state, the job is removed from the Job Service.

> When configured, all the jobs information and status changes are sent to the Kogito Data Index Service, where they can be indexed and made available by GraphQL queries.

Reviewer 2: @cristianonicolai do we have a plan to write a Data Index guide? I think it should be prioritized since we mention it here without a link. It might confuse users. :/

Author: I'll insert a sentence to mention it will be included in the future releases

==== Job Service persistence

An important configuration aspect of job service is the persistence mechanism, where all job information is stored in a database that makes this information durable upon service restarts and guarantees no information is lost.

==== PostgreSQL

> PostgreSQL is the recommended database to use with job service**.**

Reviewer 2: What's the role of this `**.**`?

Author: lol, it was added by mistake, removing it

Additionally, it provides an initialization procedure that integrates link:https://flywaydb.org[Flyway] for the database initialization. It automatically controls the database schema, in this way all tables are created by the service.

> In case you need to externally control the database schema, you can check the Flyway SQL scripts in link:{kogito_apps_url}/jobs-service/jobs-service-postgresql/src/main/resources/db/migration[migration] and apply them.

Reviewer 2: Is this `kogito_apps_url` versioned?

Author: it is pointing to main branch,  `kogito_apps_url: https://github.com/kiegroup/kogito-apps/tree/main`

You need to set the proper configuration parameters when starting job service.
The example shows how to run PostgreSQL as a Kubernetes deployment, but you can run it the way it fits in your environment, the important part is to set all the configuration parameters points to your running instance of PostgreSQL.

==== Ephemeral
Alternatively, there is an in-memory database support that does not require any external database configuration, it can be used for testing and development purposes, but it is not recommended for production, since all jobs are lost upon a service restart or failure.

=== Job Service communication

[NOTE]
====

> The Job Service does not execute a job but triggers a callback that might be an HTTP request or a Cloud Event that is managed by the configured jobs addon see(link addons config) in the workflow application.

Reviewer 2: I think we are missing the link :D

Author: yeahhh, thanks

The Job Service configuration is done through the deployment descriptor shown in the <<job-service-deploy, example>>.

==== Addon configuration in the workflow runtime

The communication from the workflow application with Job Service is done through an addon, which is responsible for publishing and consuming events related to timers.
When you run the workflow as a knative service, you must add the `kogito-addons-quarkus-jobs-knative-eventing` to your project and provide the proper configuration.

* Dependency in the `pom.xml`:

.Callback state example with timeout
[source, xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-quarkus-jobs-knative-eventing</artifactId>
</dependency>
----

* Configuration parameters:

.Callback state example with timeout
[source, properties]
----
# Events produced by kogito-addons-quarkus-jobs-knative-eventing to program the timers on the Job Service.
mp.messaging.outgoing.kogito-job-service-job-request-events.connector=quarkus-http
mp.messaging.outgoing.kogito-job-service-job-request-events.url=${K_SINK:http://localhost:8280/jobs/events}
mp.messaging.outgoing.kogito-job-service-job-request-events.method=POST

> ----

Reviewer 2: I think it's worth explaining that the `K_SINK` variable is injected by Knative Eventing and add alink here to our Knative Eventing integration guide.

Author: +1

[#callback-state]
.Example of callback state with timeout
[source,json]
----
{
 "name": "CallbackState",
 "type": "callback",
 "action": {
   "name": "callbackAction",
   "functionRef": {
     "refName": "callbackFunction",
     "arguments": {
       "input": "${\"callback-state-timeouts: \" + $WORKFLOW.instanceId + \" has executed the callbackFunction.\"}"
     }
   }
 },
 "eventRef": "callbackEvent",
 "transition": "CheckEventArrival",
 "onErrors": [
   {
     "errorRef": "callbackError",
     "transition": "FinalizeWithError"
   }
 ],
 "timeouts": {
   "eventTimeout": "PT30S"
 }
}
----

=== Switch state timeout

The switch state can be used when you need to take an action based on conditions, defined with the link:{spec_doc_url}#switch-state-event-conditions[eventConditions] property, where the workflow execution waits to make a decision depending on the events to be consumed and matched, defined through link:{spec_doc_url}#event-definition[event definition].

> In this situation, you can define an event timeout, that controls the maximum time to wait for an event to match the conditions, if this time is expired, the workflow moves to the state defined in the `defaultCondition` property of the switch state, as tou can see in the <<switch-state, example>>.

Reviewer 3: ```suggestion
In this situation, you can define an event timeout, that controls the maximum time to wait for an event to match the conditions, if this time is expired, the workflow moves to the state defined in the `defaultCondition` property of the switch state, as you can see in the <<switch-state, example>>.
```

[source, properties]
----
# Events produced by kogito-addons-quarkus-jobs-knative-eventing to program the timers on the Job Service.
mp.messaging.outgoing.kogito-job-service-job-request-events.connector=quarkus-http
mp.messaging.outgoing.kogito-job-service-job-request-events.url=${K_SINK:http://localhost:8280/jobs/events}
mp.messaging.outgoing.kogito-job-service-job-request-events.method=POST
----

[NOTE]
====
The `K_SINK` variable used in the URL configuration for the outgoing channel in the
<<workflow-application-configuration-parameters, configuration>>, is injected by Knative Eventing, more information on
xref:eventing/consume-produce-events-with-knative-eventing.adoc[Consuming and producing events on Knative Eventing].
====

[#timeout-example]
== Timeout showcase example

In the link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[serverless-workflow-timeouts-showcase] you can see an end-to-end example that contains a serverless workflow application with timeouts configured alongside Job Service running on Knative.

There are two workflows that showcase the timeouts usage in the `Callback` and `Switch` states.

=== Callback workflow

> It is a simple workflow, where once the execution reaches the callback state it waits for the event `callbackEvent`to arrive and continue the execution.

Reviewer 3: ```suggestion
It is a simple workflow, where once the execution reaches the callback state it waits for the event `callbackEvent` to arrive and continue the execution.
```

.Callback timeout workflow
image::core/callback-state-timeouts.svg[]

.Callback event
[source, json]
----
{
"name": "callbackEvent",
"source": "",
"type": "callback_event_type"
}
----

The timeout is configured with a maximum time 30 seconds to be waited by the workflow to receive the callbackEvent, in case it does not arrive in time, the execution moves, and the eventData variable remains null.
See the <<callback-state, callback state definition>>.

=== Switch workflow

The switch example is similar to the callback but once the execution reaches the state, it waits for one of the two configured events, `visaDeniedEvent` or `visaApprovedEvent`, to arrive, see the <<switch-state, switch state definition>>.

If any of the configured events arrives before the timeout is overdue, the workflow execution moves to the next state defined in the corresponding `transition`.

> If none of the events arrive before the timeout is overdue, the workflow then moves to the state defined in `defaultCondition' transition.

Reviewer 3: ```suggestion
If none of the events arrive before the timeout is overdue, the workflow then moves to the state defined in `defaultCondition` transition.
```

.Switch timeout workflow
image::core/switch-state-timeouts.svg[]

=== Running the example

To run the example you must have access to a kubernetes cluster running with Knative configured.

For simplicity, the example uses minikube, you can follow the steps described in the example's link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[readme].

[NOTE]

> ----
> All the descriptor files used to deploy the example infrastructure are present in the example.
> ----

Reviewer 3: For NOTE it needs to be
```suggestion
====
All the descriptor files used to deploy the example infrastructure are present in the example.
====
```
I believe.

trigger.eventing.knative.dev/callback-event-type-trigger-timeouts-showcase created
sinkbinding.sources.knative.dev/sb-timeouts-showcase created
----

==== Creating a workflow instance

To create a workflow you can interact with the workflow using the provided REST APIs, in the example provide a test Web UI to make it easy to test.

First, you need to get the service URL on the cluster.

.Getting the workflow service URL on the cluster
[source, shell]
----
kn service list | grep timeouts-showcase
----

.Service URL in the response, similar to this.
[source, shell]
----
NAME                      URL                                                             LATEST                          AGE     CONDITIONS   READY   REASON
timeouts-showcase         http://timeouts-showcase.default.10.105.86.217.sslip.io         timeouts-showcase-00001         3m50s   3 OK / 3     True
----

=== Using the showcase UI

The example Web UI is handy to interact with the workflow, you just need to open in the browser the URL you got from the previous step.

.Timeout workflow showcase UI
image::core/timeout-switch-wokflow-ui.png[]

> You can create new workflow instances and interact with them to complete, or simply wait the timeout to be triggered to check it working.

Reviewer 3: ```suggestion
You can create new workflow instances and interact with them to complete, or simply wait for the timeout to be triggered to check it's working.
```

[source, shell]
----
curl -X 'POST' \
'http://timeouts-showcase.default.10.105.86.217.sslip.io/callback_state_timeouts' \
-H 'accept: */*' \
-H 'Content-Type: application/json' \
-d '{
"workflowdata": {}
}'
----

* Switch

.Creating a Switch workflow with timeout
[source, shell]
----
curl -X 'POST' \
'http://timeouts-showcase.default.10.105.86.217.sslip.io/callback_state_timeouts' \
-H 'accept: */*' \
-H 'Content-Type: application/json' \
-d '{
"workflowdata": {}
}'
----

* Checking whether the workflow instance was created

.Getting the created workflow instance
[source, shell]
----
curl -X 'GET' 'http://timeouts-showcase.default.10.105.86.217.sslip.io/switch_state_timeouts'
----

The command will produce an output like this, which indicates that the process is waiting for an event to arrive.

.Response with the created instance
[source, shell]
----
[{"id":"2e8e1930-9bae-4d60-b364-6fbd61128f51","workflowdata":{}}]
----

* Checking the timeout was executed after 30 seconds:

.Getting the created workflow instance after 30 seconds
[source, shell]
----
curl -X 'GET' 'http://timeouts-showcase.default.10.105.86.217.sslip.io/switch_state_timeouts'
[]
----

> As you can see there are no active workflow instances, indicating the timout was executed and the created instance was completed.

Reviewer 3: ```suggestion
As you can see there are no active workflow instances, indicating the timeout was executed and the created instance was completed.
```
